### PR TITLE
Numpy-like interface for stack.

### LINKED ISSUE #3182: wrap numpy.stack
https://github.com/numpy/numpy/pull/5605 add it, we should make it available as a user interface and use our current op.

### PR DESCRIPTION
**Pull request by Thrandis**
Fixes #3182

### CODE REVIEW

**Comment by Thrandis** [comment]
@bouthilx : Sorry for the mess! Hope it is allright now!

File: theano/tensor/basic.py [hunk 1]
```diff
@@ -3963,16 +3963,53 @@ def shape_padright(t, n_ones=1):
...
 def stack(*tensors):
     """Insert the arguments as slices into a tensor of 1 rank greater.
 
     The size in dimension 0 of the result will be equal to the number
     of tensors passed.
 
+    Note: The interface stack(*tensors) is deprecated, you should use
+    stack(tensors, axis=0) insted.
+
+    :Parameters:
```
Comment (bouthilx) [inline theano/tensor/basic.py]: Parameters docstring format is somehow inconsistant in Theano. @lamblin Should we use 
```
"""
:Parameters: 
- `parameter_name` : type
    description
"""
```

or this

```
"""
Parameters
----------
parameter_name : type
    description
"""
```

Comment (lamblin) [inline theano/tensor/basic.py]: Ideally, the second one, but I'm not sure if it is already supported.
@abergeron, what is the status?

**Comment by bouthilx** [comment]
@Thrandis Arnaud confirmed new docstring format is supported. You can modify to this style

```python
"""
Parameters
----------
parameter_name : type
    description
"""
```

**Comment by bouthilx** [comment]
Oops. Looking at the new docstring I realized that using the function as  
```python 
stack(tensors=[A, B, C], axis=1)
``` 
would not work. We need to handle **kwargs during the deprecation phase. Sorry @Thrandis ! Please add a test case to verify this function signature.

**Comment by Thrandis** [comment]
@bouthilx there you go!

**Comment by bouthilx** [comment]
@Thrandis What if you add ```warnings.simplefilter('always', DeprecationWarning)``` at the beginning of your test? (Make sure you sure you use ```DeprecationWarning``` when you raise the message)

**Comment by Thrandis** [comment]
I was using warnings.simplefilter('always')
Le 4 sept. 2015 11:50 AM, "Xavier Bouthillier" <notifications@github.com> a
écrit :

**Comment by bouthilx** [comment]
Ok, try with DeprecationWarning. For me it didn't work without it.

**Comment by Thrandis** [comment]
I'll try that just after Aaron's class

**Comment by Thrandis** [comment]
@bouthilx I made the changes. 
The weird thing is that if the warning has been raised in the code before, then it wont't be raised again (even with `warnings.simplefilter('always', DeprecationWarning`). So I switched to the new interface for the whole test script.

**Comment by bouthilx** [comment]
oh! You should change the stack() calls to the new interface everywhere.

**Comment by Thrandis** [comment]
@bouthilx ahah damn you're right! I'll make those changes!

**Comment by Thrandis** [comment]
@bouthilx 
I added stack([a, b], axis=1) and made the change in the whole theano code!
Last thing to do is to check for those negative axis!

**Comment by Thrandis** [comment]
@bouthilx how should I handle an out-of-bounds case?

Right now, if the tensor shape is let's say 3, and we want to stack them on the axis 5, it will add a new dimension and stack them on that dimension (so the tensor will be of size 4 instead of 5), because `list.insert()` allows it.

Shall I test that `abs(axis)` is not bigger than `tensors[0].ndim + 1` and raise an error if it is?

**Comment by bouthilx** [comment]
You should test it like in numpy ([check here](https://github.com/numpy/numpy/blob/master/numpy/core/shape_base.py#L342)).

Yep if you have tensors of dimension 3, axis=5 is out of bound.

**Comment by Thrandis** [comment]
@bouthilx added negative axis. Let me know if there is other modifications to do!

**Comment by Thrandis** [comment]
@bouthilx there you go :-)

File: doc/library/tensor/basic.txt [hunk 2]
```diff
@@ -678,6 +692,26 @@ Creating Tensor
...
+    (of 1 rank greater).
+
+    :param tensors: a list or a tuple of one or more tensors of the same rank.
+    :param axis: the axis along which the tensors will be stacked.
+    :returns: A tensor such that rval[0] == tensors[0], rval[1] == tensors[1], etc.
+
+    >>> x0 = T.scalar()
+    >>> x1 = T.scalar()
+    >>> x2 = T.scalar()
+    >>> x = T.stack(x0, x1, x2)
```
Comment (bouthilx) [inline doc/library/tensor/basic.txt]: Deprecated :)

File: theano/tensor/basic.py [hunk 3]
```diff
@@ -3963,16 +3963,78 @@ def shape_padright(t, n_ones=1):
...
 
+    Note: The interface stack(*tensors) is deprecated, you should use
+    stack(tensors, axis=0) insted.
+
+    Parameters
+    ----------
+    tensors : list or tuple of tensors
+        A list of tensors to be stacked.
+    axis : int
+        The index of the new axis.
```
Comment (bouthilx) [inline theano/tensor/basic.py]: Default value is 0

**Comment by bouthilx** [comment]
Sorry @Thrandis for the two late minor fixes.

**Comment by bouthilx** [comment]
Thanks!

File: theano/tensor/basic.py [hunk 4]
```diff
@@ -3963,16 +3963,78 @@ def shape_padright(t, n_ones=1):
 
 
 @constructor
-def stack(*tensors):
+def shape_padaxis(t, axis):
+    """Reshape `t` by adding 1 at the dimension `axis`.
```
Comment (nouiz) [inline theano/tensor/basic.py]: This need an example. It is not clear on which side of the axis the 1 is added. Both option are possible.

Comment (nouiz) [inline theano/tensor/basic.py]: This also need to be documented in the file doc/library/tensor/basic.txt after shape_padright

**Pull request merged**